### DISCUSSION TITLE
Anyone get NextJS 14 (or any modern version with SWC not babel) + MikroOrm working?

### DISCUSSION THREAD

**Discussion by conspireagency** [mikro-orm/mikro-orm#5159] [General]
I've resisted using NextJS API routes for so long, but in certain cases going server less makes a lot of sense.  I want to get MikroOrm running with NextJS API routes, but it's been a HUGE pain.

The first issue I'm running into is this whole decorator metadata thing.  In dev mode everything works, but in production (built) it does not.  My tsconfig has thing like emitDecoratorMetadata enabled, so digging deeper here really scratching my head.  I was looking into SWC options, but not sure how/if NextJS allows you to modify defaults - pretty frustrating.

![Screenshot 2024-01-19 at 4 03 59 PM](https://github.com/mikro-orm/mikro-orm/assets/29992361/959a1e1e-e00a-4c1f-94c2-cad4be6b8df6)

`{
  "compilerOptions": {
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "esModuleInterop": true,
    "target": "ES2020",
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "noEmit": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "declaration": true,
    "jsx": "preserve",
    "incremental": true,
    "noUncheckedIndexedAccess": true,
    "plugins": [
      {
        "name": "next"
      }
    ],
    "paths": {
      "@/*": ["./*"]
    }
  },
  "include": [
    "next-env.d.ts",
    "**/*.ts",
    "**/*.tsx",
    ".next/types/**/*.ts",
    "**/*.d.ts"
  ],
  "exclude": ["node_modules"]
}
`

Secondly, feel like more documentation could be helpful for serverless + MikroOrm.  Most the examples / repos are years old and seemingly pretty hacky.  Looks like you need to have `import "reflect-metadata";` imported in the project, perhaps at the top of each function?  What's the best way to optimize a connection pool?  Feel like this are things that Prisma is jumping ahead with.

**Comment by B4nan** [comment]
As mentioned on slack, you need to disable class name mangling in SWC via `jsc.keepClassNames: true`. Your tsconfig is not relevant, SWC is not reading that, it has [its own config](https://swc.rs/docs/configuration/swcrc#jsckeepclassnames).

**Reply by conspireagency** [reply]
Interesting.  Problem then becomes that NextJS doesn't read ".swcrc".

Looks like as a result of [this discussion](https://github.com/vercel/next.js/discussions/37242), there may be a --no-mangling flag for "next build".  Discussion was from a while ago, so need to see if it actually does anything =0
